modules/common/condition/funcs.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2022 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate:=true
+
+package condition
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Init - init new condition list with the overall ReadyCondition set to:
+// Type: ReadyCondition, Status: Unknown, Reason, Severity and Message.
+//
+// Optional conditions list can be passed as parameter which allows to initialize
+// additional conditions at the beginning.
+func (conditions *Conditions) Init(cl *Conditions) {
+	conditions.Set(UnknownCondition(ReadyCondition, RequestedReason, ReadyInitMessage))
+
+	// add all optional conditions if no not nil
+	if cl != nil {
+		for _, c := range *cl {
+			conditions.Set(&c)
+		}
+	}
+}
+
+// Set - sets new condition on the conditions list.
+//
+// If a condition already exists, the LastTransitionTime is only updated when there is a change
+// in any of the fields: Status, Reason, Severity and Message.
+func (conditions *Conditions) Set(c *Condition) {
+	// if it is an empty condition, just return and don't set it
+	if c == nil {
+		return
+	}
+
+	// Check for the existence of a particular condition type in a list of conditions
+	// and change it only if there is a status change
+	exists := false
+	for i, existingCondition := range *conditions {
+		if existingCondition.Type == c.Type {
+			exists = true
+			if !hasSameState(&existingCondition, c) {
+				if c.LastTransitionTime.IsZero() {
+					c.LastTransitionTime = metav1.NewTime(time.Now().UTC().Truncate(time.Second))
+				}
+				(*conditions)[i] = *c
+
+				break
+			}
+			c.LastTransitionTime = existingCondition.LastTransitionTime
```

Comment:
that part is clear. We only want to update the persisted LastTransitionTime if there was a state transition. At L76 we know that there was no state transition due to the break at 74 never triggered.

My question is more about why we are updating the LastTransitionTime of the parameter c? Does the caller expects that the passed in parameter gets updated based on the persisted data?

Reply:
now I got it, sorry for not getting it sooner :) . right this is a leftover from a different previous version. this can be removed. I'll add a commit to do this which also extends the test for validate the LastTransitionTime won't change.

Reply:
ack